enhancements/etcd/storage-migration-for-etcd-encryption.md

```diff
@@ -0,0 +1,163 @@
+
+We will use the upstream `kube-storage-version-migrator` (the controller-based successor to `oc adm migrate`). While upstream usage of the `kube-storage-version-migrator` is focused on migrating to new storage versions, we should be able to explicitly trigger the `kube-storage-version-migrator` when we need to migrate the encryption key used for storage, even when the storage version itself has not changed.
+
+## Goals
+
+* Make the `StorageVersionMigration` API and its corresponding controller available.
+* Invoke the `StorageVersionMigration` API from operators requesting that storage be migrated to use new etcd storage encrpyion keys.
+
+## Non-Goals
+
+* Will not provide a higher level API to the `StorageVersionMigration` api controlled by the `kube-storage-version-migrator`.
+* Will not implement any general control loops to automatically detect and then initiate a storage migration when needed.
+
+## Proposal
+
+* Introduce `openshift/cluster-openshift-storage-migration-operator` repository.
+  * Not required for bootstrapping or re-bootstrapping.
+  * No need for static pods.
+  * Will run on cluster.
+* Implement `openshift-storage-migration-operator`. 
+* Build `kube-storage-version-migrator` image.
+  * Fork `kubernetes-sigs/kube-storage-version-migrator` to `openshift`
+  * Enable CI on `openshift/kube-storage-version-migrator` repo.
+* Utilize `StorageVersionMigration` API when etcd storage encryption keys are updated.
+
+## User Stories [optional]
+
+* Create a `openshift-storage-migration-operator`, managed by the CVO, to manage the resources needed to run the `kube-storage-version-migrator`.
+* Initiate a storage migration of resources, even if the storage version has not changed.
+* Monitor the progress of storage migration of resources.
+* Enable the `kube-apiserver-operator` to utilize the `StorageVersionMigration` API when managing etcd encryption keys.
+* Enable the `openshift-apiserver-operator` to utilize the `StorageVersionMigration` API when managing etcd encryption keys.
+* Inform the upstream `kube-storage-version-migrator` of our usage and our success/failure to gain influence for future decisions and create tests to ensure our use-cases.
+
+## Implementation Plan
+
+1. Implement `openshift-storage-migration-operator`.
+2. Have CVO manage `openshift-storage-migration-operator`.
+2. Add control loops to operators as needed.
+
+## Implementation Details/Notes/Constraints
+
+New control loop added to kube-apiserver-operator & openshift-apiserver-operator:
+
+```
+for {
+    if timeToMigrateToNewEncryptionKey {
+        for resource := range encryptedResources {
+            createStorageVersionMigration(resource)
```

Comment:
I would expect too see something about observing status updates and making the operator go progressing.

Reply:
Updated